hal/mbed_mpu_api.c

```diff
@@ -0,0 +1,120 @@
+        // Configure and enable region
+        MPU->RASR =
+                (1 << MPU_RASR_XN_Pos) |            // Execute Never
+                (3 << MPU_RASR_AP_Pos) |            // Full R/W for privileged and unprivileged
+                (1 << MPU_RASR_TEX_Pos) |           // Normal memory
+                (0 << MPU_RASR_S_Pos) |             // Non-shareable
+                (0 << MPU_RASR_C_Pos) |             // Normal memory
+                (0 << MPU_RASR_B_Pos) |             // Normal memory
+                (1 << (MPU_RASR_SRD_Pos + 0)) |     // Disable Sub-region
+                (0 << (MPU_RASR_SRD_Pos + 0)) |     // Enable Sub-region SRAM 0x20000000 - 0x3FFFFFFF
+                (1 << (MPU_RASR_SRD_Pos + 2)) |     // Disable Sub-region
+                (0 << (MPU_RASR_SRD_Pos + 0)) |     // Enable Sub-region RAM 0x60000000 - 0x7FFFFFFF
+                (0 << (MPU_RASR_SRD_Pos + 0)) |     // Enable Sub-region RAM 0x80000000 - 0x9FFFFFFF
+                (1 << (MPU_RASR_SRD_Pos + 5)) |     // Disable Sub-region
+                (1 << (MPU_RASR_SRD_Pos + 6)) |     // Disable Sub-region
+                (1 << (MPU_RASR_SRD_Pos + 7)) |     // Disable Sub-region
+                ((32 - 1) << MPU_RASR_SIZE_Pos) |   // 2**(32) == 4GB region size
+                (1 << MPU_RASR_ENABLE_Pos);         // Enable
+
+        // Select region 1
+        MPU->RNR = 1;
+        // Set address to 0
+        MPU->RBAR = 0;
+        // Configure and enable region
+        MPU->RASR =
+                (1 << MPU_RASR_XN_Pos) |            // Execute Never
+                (3 << MPU_RASR_AP_Pos) |            // Full R/W for privileged and unprivileged
+                (1 << MPU_RASR_TEX_Pos) |           // Normal memory
+                (0 << MPU_RASR_S_Pos) |             // Non-shareable
+                (0 << MPU_RASR_C_Pos) |             // Normal memory
+                (0 << MPU_RASR_B_Pos) |             // Normal memory
+                (1 << (MPU_RASR_SRD_Pos + 0)) |     // Disable Sub-region
+                (1 << (MPU_RASR_SRD_Pos + 0)) |     // Disable Sub-region
+                (1 << (MPU_RASR_SRD_Pos + 2)) |     // Disable Sub-region
+                (1 << (MPU_RASR_SRD_Pos + 0)) |     // Disable Sub-region
+                (0 << (MPU_RASR_SRD_Pos + 0)) |     // Enable Sub-region RAM 0x10000000 - 0x1FFFFFFF
+                (0 << (MPU_RASR_SRD_Pos + 5)) |     //
+                (0 << (MPU_RASR_SRD_Pos + 6)) |     //
+                (0 << (MPU_RASR_SRD_Pos + 7)) |     //
+                ((29 - 1) << MPU_RASR_SIZE_Pos) |   // 2**(29) == 0x20000000 region size
+                (1 << MPU_RASR_ENABLE_Pos);         // Enable
+    }
+
+    // Enable the MPU
+    MPU->CTRL =
+        (1 << MPU_CTRL_PRIVDEFENA_Pos) |            // Use the default memory map for unmapped addresses
+        (1 << MPU_CTRL_HFNMIENA_Pos) |              // Keep MPU turned on for faults
+        (1 << MPU_CTRL_ENABLE_Pos);                 // Enable MPU
+#endif
```

Comment:
I think we should put __DSB(), __ISB() here to ensure the memory config is completed after this point.

Reply:
done